Improve degiro v3 parsing

ikruglov opened the pull request:
## Added
This commit re-works degiroConverterV3 logic.

The main change is that the code can now consider a set of records as one instead action+tx record.

The following set of records is one sale with two transaction fee records. They can be reported as one in ghostfolio. The previous code didn't handle this.

    23-12-2020,10:03,23-12-2020,ISHARES GLOBAL CLEAN ENERGY UCITS ETF,IE00B1XNHC34,DEGIRO Transactiekosten en/of kosten van derden,,EUR,-0.79,EUR,,1d2e7a85-cf67-4325-82c7-d5994a0afb98
    23-12-2020,10:03,23-12-2020,ISHARES GLOBAL CLEAN ENERGY UCITS ETF,IE00B1XNHC34,DEGIRO Transactiekosten en/of kosten van derden,,EUR,-4.00,EUR,,1d2e7a85-cf67-4325-82c7-d5994a0afb98
    23-12-2020,10:03,23-12-2020,ISHARES GLOBAL CLEAN ENERGY UCITS ETF,IE00B1XNHC34,Verkoop 120 @ 1.202 GBX,,GBP,1442.40,GBP,,1d2e7a85-cf67-4325-82c7-d5994a0afb98

Another example is below. It has two sells and one tx record.

    11-07-2024,16:46,11-07-2024,VANGUARD FTSE ALL-WORLD UCITS ETF,IE00B3RBWM25,"Verkoop 16 @ 124,28 EUR",,EUR,1988.48,EUR,,56e1f16b-4142-4373-851d-45d0505fe12c
    11-07-2024,16:43,11-07-2024,VANGUARD FTSE ALL-WORLD UCITS ETF,IE00B3RBWM25,DEGIRO Transactiekosten en/of kosten van derden,,EUR,-1.00,EUR,,56e1f16b-4142-4373-851d-45d0505fe12c
    11-07-2024,16:43,11-07-2024,VANGUARD FTSE ALL-WORLD UCITS ETF,IE00B3RBWM25,"Verkoop 8 @ 124,28 EUR",,EUR,994.24,EUR,,56e1f16b-4142-4373-851d-45d0505fe12c

Or this

    24-03-2020,13:10,24-03-2020,VANGUARD FTSE ALL-WORLD UCITS ETF,IE00B3RBWM25,DEGIRO Transactiekosten en/of kosten van derden,,EUR,-0.26,EUR,,810d6e1d-af8e-43a4-830e-0b85bf90c818
    24-03-2020,13:10,24-03-2020,VANGUARD FTSE ALL-WORLD UCITS ETF,IE00B3RBWM25,DEGIRO Transactiekosten en/of kosten van derden,,EUR,-2.00,EUR,,810d6e1d-af8e-43a4-830e-0b85bf90c818
    24-03-2020,13:10,24-03-2020,VANGUARD FTSE ALL-WORLD UCITS ETF,IE00B3RBWM25,DEGIRO Transactiekosten en/of kosten van derden,,EUR,-0.02,EUR,,810d6e1d-af8e-43a4-830e-0b85bf90c818
    24-03-2020,13:10,24-03-2020,VANGUARD FTSE ALL-WORLD UCITS ETF,IE00B3RBWM25,"Verkoop 14 @ 61,7 EUR",,EUR,863.80,EUR,,810d6e1d-af8e-43a4-830e-0b85bf90c818
    24-03-2020,13:10,24-03-2020,VANGUARD FTSE ALL-WORLD UCITS ETF,IE00B3RBWM25,"Verkoop 1 @ 61,7 EUR",,EUR,61.70,EUR,,810d6e1d-af8e-43a4-830e-0b85bf90c818

There are also a bunch of cosmetic changes.

This is quite a significant change, so I'm looking into your feedback. I've collected a corpus of tests while debugging. I'm ready to write tests for them, but I would like your feedback to ensure you're comfortable accepting this PR.
 
## Fixes

-

## Checklist

- [ ] Added relevant changes to README (if applicable)
- [ ] Added relevant test(s)
- [ ] Updated GitVersion file and corresponding version in `package.json`

## Related issue (if applicable)

Fixes #..

ikruglov commented:
@dickwolff I just spotted https://github.com/dickwolff/Export-To-Ghostfolio/pull/146. It's still a half-measure, I believe. There are complex cases when one sale or purchase can include two or more buy/sell records and multiple transaction fee records. Please take a look at the description for examples.

dickwolff commented on src/converters/degiroConverterV3.ts:
```diff
@@ -160,33 +160,60 @@ export class DeGiroConverterV3 extends AbstractConverter {
+
+        // If there was no match by orderId, and there was no orderId present on the current record, look ahead one record to find a match by ISIN + Product + Date/Time.
+        if (!matchingRecords && !record.orderId) {
+          matchingRecords = this.findMatchByIsinSameDateTime(record, deGiroRecords.slice(idx + 1, idx + 2));
```
I left out time previously because this can actually differ between records and then produce an unwanted result.

ikruglov replied:
Do you have examples? The previous logic did an unbounded search, which was too much. We may remove time (but keep date) or even remove date/time but then need to limit search window somehow (number of lines?)

dickwolff replied:
I recently fixed a bug, and the reporter had a dividend set which had a different time.

```
01-08-2024,07:40,31-07-2024,OXFORD SQUARE CAPITAL CORP,US69181V1070,Impôts sur dividende,,USD,-0.53,USD,2.97,
01-08-2024,07:39,31-07-2024,OXFORD SQUARE CAPITAL CORP,US69181V1070,Dividende,,USD,3.50,USD,3.50,
```

ikruglov replied:
> I recently fixed a bug, and the reporter had a dividend set which had a different time.
> 
> ```
> 01-08-2024,07:40,31-07-2024,OXFORD SQUARE CAPITAL CORP,US69181V1070,Impôts sur dividende,,USD,-0.53,USD,2.97,
> 01-08-2024,07:39,31-07-2024,OXFORD SQUARE CAPITAL CORP,US69181V1070,Dividende,,USD,3.50,USD,3.50,
> ```

I'd suggest to limit by date (not time) then

dickwolff commented on src/models/degiroRecord.ts:
```diff
@@ -8,5 +8,17 @@ export class DeGiroRecord {
+    public getAmount(): Number { return parseFloat(this.amount.replace(",", ".")); }
+    public getAbsoluteAmount(): Number { return Math.abs(this.getAmount()); }
+
+    public getFX(): Number { return parseFloat(this.fx.replace(",", ".")); }
```
In the time I have maintained the tool I haven’t found fx to be relevant until now. Why would it be?

ikruglov replied:
yeah, I think I added it for completeness. I can remove if you prefer.

dickwolff replied:
Yes please

dickwolff commented on src/converters/degiroConverterV3.ts:
```diff
@@ -21,6 +22,12 @@ export class DeGiroConverterV3 extends AbstractConverter {
    * @inheritdoc
    */
   public processFileContents(input: string, successCallback: any, errorCallback: any): void {
+    // my exported file contained many broken records like this. It's easy to recover them. So, why not?!
```
What exactly is broken about this?

ikruglov replied:
check the comment below. The record is split between two lines with many commas in the middle.

dickwolff commented on src/converters/degiroConverterV3.ts:
```diff
@@ -62,90 +69,83 @@ export class DeGiroConverterV3 extends AbstractConverter {
+         * since buy/sell txs are in EUR.
+         *
+         * This hack allows to buypass all internal heuristic and hardcode correct results.
+         * Format: IE00B02KXK85=FXAC.AS,foo=bar
```
I have seen a similar thing with the isin of VWRL resulting in VWRL.L instead of VWRL.AS (see comment on line 92 on main).

I think you’re on the right path to a solution, but I think it can be improved upon. I’d then rather use a simple text file with mappings which you as a user can provide.

I have created #149 for this to make this separate of this change. Could you revert any change regarding this mapping override, so the this PR is moreso a better fix for the transaction matching?

ikruglov replied:
Do you have ready code? If not, I can write it since I need it. Do you have any particular preferences implementation wise?

dickwolff replied:
No I have no code yet, but your suggestion gave me an idea on how to do it.

You can give it a go, but preferably separate from, and after this change has been merged

ikruglov replied:
> No I have no code yet, but your suggestion gave me an idea on how to do it.
> 
> You can give it a go, but preferably separate from, and after this change has been merged

so, just a plain file with `foo=bar` pairs on each line?

dickwolff replied:
That was basically what I was going for. Then read it when present, and also provide a file volume mount in Docker.

But as stated, I’d prefer it separate from this PR

dickwolff commented on package.json:
```diff
@@ -32,7 +32,8 @@
     "tsx": "^4.19.2",
-    "yahoo-finance2": "^2.13.3"
+    "yahoo-finance2": "^2.13.3",
+    "ts-md5": "^1.3.1"
```
You can also hash with plain node, not requiring an additional library. Like below:

```ts
import * as crypto from “crypto”;
var name = 'braitsch';
var hash crypto.createHash('md5').update(name).digest('hex');
```

ikruglov replied:
ah, ok. I'm not an experienced typescript coder. It's my first ts ever. So, I went for an approach which googled.

dickwolff commented on src/converters/degiroConverterV3.ts:
```diff
@@ -62,90 +69,83 @@ export class DeGiroConverterV3 extends AbstractConverter {
-          bar1.increment();
-          continue;
+        // Check if the current record was already processed.
+        const recordHash = this.hashRecord(record);
```
This might be a better solution. However I need to look into this better on my pc where I have a better overview of the whole change set.

ikruglov replied:
sure, let me know

dickwolff reviewed (changes requested):
Thanks for the suggestions on how to improve the mapping for DEGIRO. I have done an initial review of the changes. Overall I think you’re on the right path, but I have some remarks, questions or suggestions. 

Please note that I have reviewed it on my iPad, so I might not have the full picture of all the changes. I will look into it some more later when I can get on a pc.

dickwolff commented on src/models/degiroRecord.ts:
```diff
@@ -8,5 +8,17 @@ export class DeGiroRecord {
+    balance_currency: string; // not used, but improtant for hashing
+    balance: string; //// not used, but improtant for hashing
     orderId: string;
+
```
The test are also breaking due to you using Number instead of number. 

See https://github.com/dickwolff/Export-To-Ghostfolio/actions/runs/12614130746/job/35162022748

ikruglov replied:
no problem. Once again. I had never used ts before, so I didn't know the difference. I'll switch to `number`

ikruglov commented:
btw, do you prefer rebase + force-push or extra commits?

dickwolff commented:
I recently tried to switch to rebasing, so please

ikruglov commented:
@dickwolff I fixed all your suggestions and added tests. But I didn't figure out how to update Yahoo Finance mock file to include info about new ISINs. How do I do it?

ikruglov commented:
@dickwolff I just realized that the import logic seriously messes up with currencies. It just replaces record currency with looked-up-from-yahoo currency without any proper conversation.

I understand why, but it affects results. In my case, with INRG.L, quite seriously.

Asked question: https://github.com/ghostfolio/ghostfolio/issues/4139#issuecomment-2571660076

What was your thinking when you wrote it?

dickwolff commented:
> @dickwolff I fixed all your suggestions and added tests. But I didn't figure out how to update Yahoo Finance mock file to include info about new ISINs. How do I do it?

See the wiki for instructions how to do this.

----

> @dickwolff I just realized that the import logic seriously messes up with currencies. It just replaces record currency with looked-up-from-yahoo currency without any proper conversation.
> 
> I understand why, but it affects results. In my case, with INRG.L, quite seriously.
> 
> Asked question: [ghostfolio/ghostfolio#4139 (comment)](https://github.com/ghostfolio/ghostfolio/issues/4139#issuecomment-2571660076)
> 
> What was your thinking when you wrote it?

This was a fix for GBP to GBp matching. I have since made some changes in a branch to revert that change, in response to #129. I have not merged that yet due to something I wanted to verify with the reporter first. But I think I will do that merge tomorrow, so you can continue working on that.

jonathanR79 commented:
Some other samples of multi line transaction record that are currentlly not really well processed:

```
10-12-2020,17:54,10-12-2020,ADR ON AMBEV,US02319V1035,Operation de change - Crédit,1.2105,USD,311.00,USD,0.00,a08ddbcc-73da-4250-ac23-yyyyyyyyy
10-12-2020,17:54,10-12-2020,ADR ON AMBEV,US02319V1035,Opération de change - Débit,,EUR,-256.92,EUR,754.35,a08ddbcc-73da-4250-ac23-yyyyyyyyy
10-12-2020,17:54,10-12-2020,ADR ON AMBEV,US02319V1035,Taxe sur les Opérations Boursières belge (TOB),,EUR,-0.90,EUR,1011.27,a08ddbcc-73da-4250-ac23-yyyyyyyyy
10-12-2020,17:54,10-12-2020,ADR ON AMBEV,US02319V1035,Frais DEGIRO de courtage et/ou de parties tierces,,EUR,-0.33,EUR,1012.17,a08ddbcc-73da-4250-ac23-yyyyyyyyy
10-12-2020,17:54,10-12-2020,ADR ON AMBEV,US02319V1035,Frais DEGIRO de courtage et/ou de parties tierces,,EUR,-0.50,EUR,1012.50,a08ddbcc-73da-4250-ac23-yyyyyyyyy
10-12-2020,17:54,10-12-2020,ADR ON AMBEV,US02319V1035,"Achat 100 ADR on Ambev@3,11 USD (US02319V1035)",,USD,-311.00,USD,-311.00,a08ddbcc-73da-4250-ac23-yyyyyyyyy
```

```
04-06-2020,12:21,04-06-2020,QUEST FOR GROWTH,BE0003730448,Taxe sur les Opérations Boursières belge (TOB),,EUR,-0.30,EUR,506.28,b5c849a1-baef-473e-bdd7-yyyyyyyyy
04-06-2020,12:21,04-06-2020,QUEST FOR GROWTH,BE0003730448,Frais DEGIRO de courtage et/ou de parties tierces,,EUR,-0.10,EUR,506.58,b5c849a1-baef-473e-bdd7-yyyyyyyyy
04-06-2020,12:21,04-06-2020,QUEST FOR GROWTH,BE0003730448,Taxe sur les Opérations Boursières belge (TOB),,EUR,-0.30,EUR,506.68,b5c849a1-baef-473e-bdd7-yyyyyyyyy
04-06-2020,12:21,04-06-2020,QUEST FOR GROWTH,BE0003730448,Frais DEGIRO de courtage et/ou de parties tierces,,EUR,-0.10,EUR,506.98,b5c849a1-baef-473e-bdd7-yyyyyyyyy
04-06-2020,12:21,04-06-2020,QUEST FOR GROWTH,BE0003730448,"Achat 50 Quest For Growth@4,98 EUR (BE0003730448)",,EUR,-249.00,EUR,507.08,b5c849a1-baef-473e-bdd7-yyyyyyyyy
04-06-2020,12:21,04-06-2020,QUEST FOR GROWTH,BE0003730448,"Achat 50 Quest For Growth@4,97 EUR (BE0003730448)",,EUR,-248.50,EUR,756.08,b5c849a1-baef-473e-bdd7-yyyyyyyyy
```

dickwolff commented:
@jonathanR79 Thanks for providing these. Of the first set, the first records should be ignored. I think that currently is already the case. Of the second set, would you want there to be 2 buy records, or is it fine to combine all these 6 lines into one activity for Ghostfolio?

jonathanR79 commented:
> @jonathanR79 Thanks for providing these. Of the first set, the first records should be ignored. I think that currently is already the case. Of the second set, would you want there to be 2 buy records, or is it fine to combine all these 6 lines into one activity for Ghostfolio?

On the first, it's detected as a divident fo rnow.

On the second, the purchase price is different. So , ideally it should be two different records. A single entry with the average of the 2 should work as well but I'm not sure what is the easiest.

ikruglov commented:
@dickwolff I've pushed an improved version. Unfortunately, I'm unable to test it locally. For some reason, ghostfolio tells me, "Don't have a currency rate from EUR to USD." I'm unable to work around it for now.

ikruglov commented:
> On the second, the purchase price is different. So , ideally it should be two different records. A single entry with the average of the 2 should work as well but I'm not sure what is the easiest.

@jonathanR79 this code will do two records. One with combination of TX costs, second one without one tx cost.

dickwolff commented:
> @dickwolff I've pushed an improved version. Unfortunately, I'm unable to test it locally. For some reason, ghostfolio tells me, "Don't have a currency rate from EUR to USD." I'm unable to work around it for now.

You can should be able to this via Admin Control. Click the `+ Add Currency`-button and enter `EUR`. Then you should have the exchange rate like `1 USD = 0,9745 EUR` (on time of writing).

ikruglov commented:
> > @dickwolff I've pushed an improved version. Unfortunately, I'm unable to test it locally. For some reason, ghostfolio tells me, "Don't have a currency rate from EUR to USD." I'm unable to work around it for now.
> 
> You should be able to this via Admin Control. Click the `+ Add Currency`-button and enter `EUR`. Then you should have the exchange rate like `1 USD = 0,9745 EUR` (on time of writing).
> 
> I will hopefully have some time this weekend to go through your changes and leave review.

I tried that. It doesn't help. I'm kinda stuck. I see that ghostfolio extracts and adds historical data for USDEUR, but it fails with the opposite conversation: EURUSD. I don't know/see a way to import it.

ikruglov commented:
oh, I figured it out. It's a bug in ghostfolio. I'm preparing a PR.

ikruglov commented:
submitted https://github.com/ghostfolio/ghostfolio/pull/4187

dickwolff commented on src/converters/degiroConverterV3.test.ts:
```diff
@@ -35,7 +35,7 @@ describe("degiroConverterV3", () => {
       expect(actualExport).toBeTruthy();
       expect(actualExport.activities.length).toBeGreaterThan(0);
-      expect(actualExport.activities.length).toBe(26);
+      expect(actualExport.activities.length).toBe(43);
```
I see the [tests are currently failing](https://github.com/dickwolff/Export-To-Ghostfolio/actions/runs/12699976144/job/35632606126). V2 converter is missing 1 (as V2 and V3 tests are sharing the same input file for testing), and V3 only got 31 instead of expecting the 43 you put in the test. Can you take a look at this and make changes?

ikruglov replied:
@dickwolff I need help with tests:
1. to update Yahoo Finance information. I think the instructions from the wiki are not complete. I got a missing import of a Mock object, which I didn't figure out how to import. Note that I'm a first-time user of Typescript
2. How do I print debug statements (to stdout/stderr) in tests? I'm getting unexpected behavior, which I need to debug.

dickwolff replied:
Sure! Hope this helps:

1. If you use an IDE like VSCode, you can to `CTRL + .` (dot) and then import all usings. 
2. In the tests, console statements are supressed. You can comment the whole `beforeEach(() => {` function, then you'll get the console log statements in your console. For debugging purposes I usually add `console.log(...)` statements on places that could help me debug.

dickwolff reviewed (changes requested):
I just looked at the changes and they seem to make sense overall. I might do some testing and code cleanup myself after merging, but the idea is sound. Great work!

Finally, can you bump the versions to 0.23.0?

dickwolff commented:
@ikruglov If you agree, you can change this converter to v4 (and copy back the old converter as v3). I can merge the changes to main and I can finalize it there, as well take a look at fixing the tests. 

Let me know what you think!

ikruglov commented:
Hey. Go ahead. I've decided to go with another option to manage portfolio so I don't think I will have more time for it.

dickwolff commented:
Sure thing! I’ll try to make the rename change on your branch, otherwise I’ll copy the changes. 

Out of curiosity, what made you choose another portfolio management tool (and which did you land on?)

ikruglov commented:
Fixing problems which I care about started taking too much time.
